runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java

```diff
@@ -325,83 +326,103 @@ protected void initChannel(@Nonnull Channel ch) throws Exception {
+            connectionFuture = new CompletableFuture<>();
+            // Exceptionally complete all requests that were waiting for a completion.
             outstandingRequests.forEach((reqId, reqCompletableFuture) -> {
-                reqCompletableFuture.completeExceptionally(new NetworkException("Disconnected",
-                    node));
-                outstandingRequests.remove(reqId);
+                reqCompletableFuture.completeExceptionally(
+                        new NetworkException("Disconnected", node));
+            // And also remove them.
+            outstandingRequests.remove(reqId);
             });
+            // If we aren't shutdown, reconnect.
             if (!shutdown) {
-                log.trace("Disconnected, reconnecting...");
-                while (!shutdown) {
-                    try {
-                        connectChannel(b);
-                        return;
-                    } catch (SocketException e) {
-                        log.warn("Exception while reconnecting, retry in {} ms", timeoutRetry, e);
-                        Sleep.MILLISECONDS.sleepUninterruptibly(timeoutRetry);
-                    }
+                log.info("addReconnectionOnCloseFuture[{}]: reconnecting", node);
+                // Asynchronously connect again.
+                connectAsync(b);
+            }
+        });
+    }
+
+    /** Connect to a remote server asynchronously.
+     *
+     * @param b         The channel boostrap to use
+     * @return          A {@link ChannelFuture} which is c
+     */
+    private ChannelFuture connectAsync(@Nonnull Bootstrap b) {
+        // If shutdown, return a ChannelFuture that is exceptionally completed.
+        if (shutdown) {
+            return new DefaultChannelPromise(channel, GlobalEventExecutor.INSTANCE)
+                .setFailure(new ShutdownException("Runtime already shutdown!"));
+        }
+        // Use the bootstrap to create a new channel.
+        ChannelFuture f = b.connect(node.getHost(), node.getPort());
+        f.addListener((ChannelFuture future) -> {
+            if (future.isSuccess()) {
+                // When the channel connection succeeds, set this channel as active
+                channel = future.channel();
+                // And register a future to reconnect in case we get disconnected
+                addReconnectionOnCloseFuture(channel, b);
+                log.info("connectAsync[{}]: Channel connected.", node);
+            } else {
+                // Otherwise, the connection failed. If we're not shutdown, try reconnecting after
+                // a sleep period.
+                if (!shutdown) {
+                    log.info("connectAsync[{}]: Channel connection failed, reconnecting...", node);
+                    Sleep.sleepUninterruptibly(parameters.getConnectionRetryRate());
+                    // Call connect, which will retry the call again.
+                    connectAsync(b);
```

Comment:
This is recursive ? On node failure is it possible that we can overflow the stack ? I am not comfortable with this as this does not seem tail recursive either.

Reply:
It is not recursive. The call is in a future, unlike before.

Reply:
but connectAsync(b) is being called in connectAsync(b). if there is failure what happens ?